qutip/entropy.py

```diff
@@ -223,45 +224,79 @@ def entropy_mutual(rho, selA, selB, base=e, sparse=False):
     return out
 
 
-def _entropy_relative(rho, sigma, base=e, sparse=False):
+def entropy_relative(rho, sigma, base=e, sparse=False, tol=1e-12):
     """
-    ****NEEDS TO BE WORKED ON****
-
     Calculates the relative entropy S(rho||sigma) between two density
     matrices.
-
     Parameters
     ----------
-    rho : qobj
-        First density matrix.
-    sigma : qobj
-        Second density matrix.
+    rho : :class:`qutip.Qobj`
+        First density matrix (or ket which will be converted to a density
+        matrix).
+    sigma : :class:`qutip.Qobj`
+        Second density matrix (or ket which will be converted to a density
+        matrix).
     base : {e,2}
-        Base of logarithm.
+        Base of logarithm. Defaults to e.
+    sparse : bool
+        Flag to use sparse solver when determining the eigenvectors
+        of the density matrices. Defaults to False.
+    tol : float
+        Tolerance to use to detect 0 eigenvalues or dot producted between
+        eigenvectors. Defaults to 1e-12.
 
     Returns
     -------
     rel_ent : float
         Value of relative entropy.
 
+    References
+    ----------
+
+    See Nielsen & Chuang, "Quantum Computation and Quantum Information",
+    Section 11.3.1, pg. 511 for a detailed explanation of quantum relative
+    entropy.
     """
-    if rho.type != 'oper' or sigma.type != 'oper':
-        raise TypeError("Inputs must be density matrices..")
-    # sigma terms
-    svals = sp_eigs(sigma.data, sigma.isherm, vecs=False, sparse=sparse)
-    snzvals = svals[svals != 0]
+    if rho.isket:
+        rho = ket2dm(rho)
+    if sigma.isket:
+        sigma = ket2dm(sigma)
+    if not rho.isoper or not sigma.isoper:
+        raise TypeError("Inputs must be density matrices.")
+    if rho.dims != sigma.dims:
+        raise ValueError("Inputs must have the same shape and dims.")
     if base == 2:
-        slogvals = log2(snzvals)
+        log_base = log2
     elif base == e:
-        slogvals = log(snzvals)
+        log_base = log
     else:
         raise ValueError("Base must be 2 or e.")
-    # rho terms
-    rvals = sp_eigs(rho.data, rho.isherm, vecs=False, sparse=sparse)
-    rnzvals = rvals[rvals != 0]
-    # calculate tr(rho*log sigma)
-    rel_trace = float(real(sum(rnzvals * slogvals)))
-    return -entropy_vn(rho, base, sparse) - rel_trace
+    # S(rho || sigma) = sum_i(p_i log p_i) - sum_ij(p_i P_ij log q_i)
+    #
+    # S is +inf if the kernel of sigma (i.e. svecs[svals == 0]) has non-trivial
+    # intersection with the support of rho (i.e. rvecs[rvals != 0]).
+    rvals, rvecs = sp_eigs(rho.data, rho.isherm, vecs=True, sparse=sparse)
+    if any(imag(rvals) >= tol):
+        raise ValueError("Input rho has non-real eigenvalues.")
```

Comment:
What if there are negative imaginary components?

(Also, incredibly minor nit, but one of my pet peeves - the pre-existing style of this file [import everything into a global namespace] masks the fact that the `any` being called is the Python iterator built-in, not the vectorised `np.any`.)

Reply:
(Although `np.any` doesn't short-circuit.)

Reply:
Good catch. Fixed.

Re `any` vs `np.any` -- yeah, the style of this file has been annoying me too. I have been making myself avoid fixing that in this PR. :)

Reply:
Haha, not fixing stuff that's "not my style" is definitely something I have to force myself to do as well!